contracts/Relay.sol

```diff
@@ -0,0 +1,431 @@
+    /// @dev                      We don't check this AT ALL really. Don't use relays with bad genesis
+    /// @param  _genesisHeader    The starting header
+    /// @param  _height           The starting height
+    constructor(bytes memory _genesisHeader, uint256 _height, bytes32 _periodStart) public {
+        require(_genesisHeader.length == 80, "Stop being dumb");
+        bytes32 _genesisDigest = _genesisHeader.hash256();
+        relayGenesis = _genesisDigest;
+        bestKnownDigest = _genesisDigest;
+        lastReorgCommonAncestor = _genesisDigest;
+        blockHeight[_genesisDigest] = _height;
+        blockHeight[_periodStart] = _height.sub(_height % 2016);
+    }
+
+    /// @notice             Adds headers to storage after validating
+    /// @dev                We check integrity and consistency of the header chain
+    /// @param  _anchor     The header immediately preceeding the new chain
+    /// @param  _headers    A tightly-packed list of new 80-byte Bitcoin headers to record
+    /// @param  _internal   True if called internally from addHeadersWithRetarget, false otherwise
+    /// @return             True if successfully written, error otherwise
+    function _addHeaders(bytes memory _anchor, bytes memory _headers, bool _internal) internal returns (bool) {
+        uint256 _height;
+        bytes memory _header;
+        bytes32 _currentDigest;
+        bytes32 _previousDigest = _anchor.hash256();
+
+        uint256 _target = _headers.slice(0, 80).extractTarget();
+        uint256 _anchorHeight = _findHeight(_previousDigest);  /* NB: errors if unknown */
+
+        require(
+            _internal || _anchor.extractTarget() == _target,
+            "Unexpected retarget on external call");
+        require(_headers.length % 80 == 0, "Header array length must be divisible by 80");
+        require(_headers.length / 80 >= 5, "Must supply at least 5 headers");
+
+        /*
+        NB:
+        1. check that the header has sufficient work
+        2. check that headers are in a coherent chain (no retargets, hash links good)
+        3. Store the block connection
+        4. Store the height
+        */
+        for (uint256 i = 0; i < _headers.length / 80; i = i.add(1)) {
+            _header = _headers.slice(i.mul(80), 80);
+            _height = _anchorHeight.add(i).add(1);
+            _currentDigest = _header.hash256();
+
+            /*
+            NB:
+            if the block is already authenticated, we don't need to a work check
+            Or write it anything to state.
+            This saves gas
+            */
+            if (previousBlock[_currentDigest] == bytes32(0)) {
+                require(
+                    abi.encodePacked(_currentDigest).reverseEndianness().bytesToUint() <= _target,
+                    "Header work is insufficient");
+                previousBlock[_currentDigest] = _previousDigest;
+                if (_height % 4 == 0) {
```

Comment:
Can't you store just the Nth block when providing with a header chain of length N? How is 4 different from any other parameter?

Reply:
I'm really on the fence about allowing this to be parameterized. Ideally, this contract is a chain-wide singleton. 4 seems like a reasonable balance between saving money on writes and keeping `findHeight` calls cheap

4 means 5,000 gas per header (amortized)
and lookups cost >=600 (3 `SLOAD`)

8 would mean 2,500 gas per header (amortized)
and lookups costing  >=1,400 (7 `SLOAD`)

Considering we expect to do >1 lookup per 4 headers, (1 for extending the chain, and some amount for marking new tips), it seems like 8 is too sparse